pat/consumer_tests/step_6.py

```diff
@@ -0,0 +1,565 @@
+                extra={'step': step},
+            )
+            test_results.append(False)
+            continue
+        if len(operation_result.report_parts) != len(expected_order):
+            logger.error(
+                'The SetString operation with the handle %s finished with %d report parts, but expected exactly %d',
+                operation.Handle,
+                len(operation_result.report_parts),
+                len(expected_order),
+                extra={'step': step},
+            )
+            test_results.append(False)
+            continue
+        for report in (
+            report
+            for report in operation_result.report_parts
+            if report.InvocationInfo.InvocationState not in expected_order
+        ):
+            logger.error(
+                'The SetString operation with the handle %s returned unexpected invocation state %s with error %s: %s',
+                operation.Handle,
+                report.InvocationInfo.InvocationState,
+                report.InvocationInfo.InvocationError,
+                report.InvocationInfo.InvocationErrorMessage,
+                extra={'step': step},
+            )
+            test_results.append(False)
+        for expected, actual in zip(
+            expected_order,
+            (p.InvocationInfo.InvocationState for p in operation_result.report_parts),
+            strict=True,
+        ):
+            if expected == actual:
+                logger.info(
+                    'The answer of the SetString operation with the handle %s operation is "%s" as expected',
+                    operation.Handle,
+                    expected,
+                    extra={'step': step},
+                )
+                test_results.append(True)
+            else:
+                logger.error(
+                    'The answer of the SetString operation with the handle %s operation is "%s", but expected "%s"',
+                    operation.Handle,
+                    actual,
+                    expected,
+                    extra={'step': step},
+                )
+                test_results.append(False)
+    return any(test_results) and all(test_results)
+
+
+def _observe_metric_value_change(
+    expected_values_by_handle: dict[str, typing.Any],
+    observed_handles: set[str],
+    completion_event: threading.Event,
+    step: str,
+) -> typing.Callable[[dict[str, statecontainers.AbstractMetricStateContainer]], None]:
+    """Return an observer that marks when all expected metric values were reported."""
+    expected_handles = set(expected_values_by_handle)
+
+    def _observer(metrics_by_handle: dict[str, statecontainers.AbstractMetricStateContainer]) -> None:
+        if completion_event.is_set():
+            return
+        for descriptor_handle, state in metrics_by_handle.items():
+            if descriptor_handle not in expected_handles:
+                continue
+            metric_value = getattr(state.MetricValue, 'Value', None)
```

Comment:
would suggest to do:
`metric_value = state.MetricValue.Value if state.MetricValue else None`

using `getattr` is not the best way to do in my opninion